File: Source/Common/Microsoft.Deployment.Common/Actions/ActionResponse.cs
using System;
using Microsoft.Deployment.Common.ErrorCode;
using Microsoft.Deployment.Common.Helpers;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Microsoft.Deployment.Common.Actions
{
    public class ActionResponse
    {
        // Status of the request
        [JsonConverter(typeof(StringEnumConverterLower))]
        public ActionStatus Status;

        public JObject Response;

        public Exception ExceptionCaught;

        public string LogLocation { get; set; }

        public string Message { get; set; }

        public string FriendlyMessageCode { get; set; }

        public string FriendlyErrorMessage
        {
            get
            {
                if (this.FriendlyMessageCode == null)
                {
                    return string.Empty;
                }

                 return ErrorUtility.GetErrorCode(this.FriendlyMessageCode);
            }
        }

        public string AdditionalDetailsErrorMessage { get; set; }


        // Used to Serialize and Deserialize
        public ActionResponse()
        {

        }

        public ActionResponse(ActionStatus status) : this(status, JsonUtility.GetEmptyJObject())
        {
        }

        public ActionResponse(ActionStatus status, JObject response)
        {
            if (status == ActionStatus.Failure || status == ActionStatus.FailureExpected)
            {
                this.FriendlyMessageCode = DefaultErrorCodes.DefaultErrorCode;
            }
            this.Status = status;
            this.Response = response;
        }

        public ActionResponse(ActionStatus status, string response)
        {
            JObject obj = new JObject();

            if (status == ActionStatus.Failure)
            {
                this.FriendlyMessageCode = DefaultErrorCodes.DefaultErrorCode;
            }
            if (!string.IsNullOrEmpty(response))
            {
                obj = JsonUtility.GetJsonObjectFromJsonString(response);
            }

            this.Status = status;
            this.Response = obj;
        }

        public ActionResponse(ActionStatus status, object response)
        {
            if (status == ActionStatus.Failure)
            {
                this.FriendlyMessageCode = DefaultErrorCodes.DefaultErrorCode;
            }

            this.Status = status;
            this.Response = JsonUtility.GetJObjectFromObject(response);
        }

        public ActionResponse(ActionStatus status, object response, string friendlyMessageCode)
        {
            if (status == ActionStatus.Failure)
            {
                this.FriendlyMessageCode = friendlyMessageCode;
            }

            this.Status = status;
            this.Response = JsonUtility.GetJObjectFromObject(response);
        }

        public ActionResponse(ActionStatus status, JObject response, Exception exception,
            string friendlyMessageCode, string additionaldetails)
        {
            this.Status = status;
            this.Response = response;
            this.ExceptionCaught = exception;
            this.FriendlyMessageCode = friendlyMessageCode;
            this.AdditionalDetailsErrorMessage = additionaldetails;
        }

        public ActionResponse(ActionStatus status, JObject response, Exception exception,
            string friendlyMessageCode)
        {
            this.Status = status;
            this.Response = response;
            this.ExceptionCaught = exception;
            this.FriendlyMessageCode = friendlyMessageCode;
            this.AdditionalDetailsErrorMessage = GetInnerExceptionText(exception);
        }

        private static string GetInnerExceptionText(Exception exception)
        {
            if (exception == null)
                return string.Empty;

            Exception loopException = exception;
            while (loopException.InnerException != null)
            {
                loopException = loopException.InnerException;
            }

            return loopException.Message;
        }
    }
}

File: Source/SiteCommon/Web/services/actionresponse.ts
export class ActionResponse {
    response: any;
    isSuccess: boolean;
    errorMessage: any;
    logLocation: string;
    responseStatus: ActionStatus;
    friendlyErrorMessage: string;
    friendlyMessageCode: string;
    additionalDetailsErrorMessage: string;

    constructor(response) {
        let responseParsed: any = JSON.parse(response);

        this.isSuccess = !responseParsed.Status.startsWith('failure');

        if (responseParsed.Status === 'failure') {
            this.responseStatus = ActionStatus.Failure;
        }

        if (responseParsed.Status === 'failureexpected') {
            this.responseStatus = ActionStatus.FailureExpected;
        }

        if (responseParsed.Status === 'batchnostate') {
            this.responseStatus = ActionStatus.BatchNoState;
        }

        if (responseParsed.Status === 'batchwithstate') {
            this.responseStatus = ActionStatus.BatchWithState;
        }

        if (responseParsed.Status === 'userinteractionrequired') {
            this.responseStatus = ActionStatus.UserInteractionRequired;
        }

        if (responseParsed.Status === 'success') {
            this.responseStatus = ActionStatus.Success;
        }

        this.response = responseParsed.Response;
        this.friendlyErrorMessage = responseParsed.FriendlyErrorMessage;
        this.friendlyMessageCode = responseParsed.FriendlyMessageCode;
        this.additionalDetailsErrorMessage = responseParsed.AdditionalDetailsErrorMessage;
        this.logLocation = responseParsed.LogLocation;
    }
}

export enum ActionStatus {
    Failure,
    FailureExpected,
    BatchNoState,
    BatchWithState,
    UserInteractionRequired,
    Success,
}